extensions/amp-ad-network-cloudflare-impl/0.1/amp-ad-network-cloudflare-impl.js

```diff
@@ -64,16 +56,43 @@ export class AmpAdNetworkCloudflareImpl extends AmpA4A {
 
   /** @override */
   getAdUrl() {
-    // You can optionally use data from this context to augment the URL,
-    // to pass data back to the network, to aid in ad selection.  For example:
-    //   const rect = this.getIntersectionElementLayoutBox();
-    // can extract the layout size and be used to pass layout hints to the server.
+    const rect = this.getIntersectionElementLayoutBox();
+    const el = this.element;
+
+    const network = el.getAttribute('data-cf-network');
+    const base = NETWORKS[network].base;
+
+    // handle variable replacements
+    function replacements(str) {
+      if (!str) {
+        return str;
+      }
+
+      // allow injection of width and height as parameters
+      str = str.replace(/{width}/g, (rect.width || 0).toString());
+      str = str.replace(/{height}/g, (rect.height || 0).toString());
+
+      return str;
+    }
 
-    // If you don't want to use relative URL generation from your base url,
-    // be sure that this only generates urls to valid hosts, since it could
-    // allow cookie leaks, etc.
+    let url = base + '/_a4a' + el.getAttribute('src');
+
+    // encode for safety
+    url = encodeURI(replacements(url));
+
+    // include other data attributes as query parameters
+    let pre = url.match(/\?/) ? '&' : '?';
+    for (let i = 0; i < el.attributes.length; i++) {
```

Comment:
If you want that, you probably shouldn't have the keys be HTML attributes; those don't preserve case and I don't think they allow all the same characters that URL parameters do. This might be tricky enough to get right as to be best left off this PR and added later.